File: horizons_ts/src/agents.ts
import { HorizonsClient, HorizonsError } from "./client";
import { AgentRunResult } from "./types";

export class AgentsAPI {
  constructor(private client: HorizonsClient) {}

  async run(agent_id: string, inputs?: unknown, project_id?: string): Promise<AgentRunResult> {
    const body: Record<string, unknown> = { agent_id };
    if (inputs !== undefined) body.inputs = inputs;
    if (project_id) body.project_id = project_id;
    const resp = await this.client.request<{ result: AgentRunResult }>("/api/v1/agents/run", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });
    return resp.result;
  }

  async *chatStream(agent_id: string, inputs?: unknown, project_id?: string): AsyncGenerator<Record<string, unknown>> {
    const body: Record<string, unknown> = { agent_id };
    if (inputs !== undefined) body.inputs = inputs;
    if (project_id) body.project_id = project_id;

    const resp = await fetch(`${this.client.baseUrl}/api/v1/agents/chat`, {
      method: "POST",
      headers: { ...this.client["headers"]({ "content-type": "application/json", accept: "text/event-stream" }) },
      body: JSON.stringify(body),
    });
    if (!resp.ok || !resp.body) {
      throw new HorizonsError(`chat stream failed: ${resp.statusText}`, resp.status);
    }
    const reader = resp.body.getReader();
    let buffer = "";
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += new TextDecoder().decode(value);
      let sepIndex;
      while ((sepIndex = buffer.indexOf("\n\n")) >= 0) {
        const chunk = buffer.slice(0, sepIndex);
        buffer = buffer.slice(sepIndex + 2);
        const dataLine = chunk.split("\n").find((line) => line.startsWith("data:"));
        if (!dataLine) continue;
        const jsonStr = dataLine.replace("data:", "").trim();
        if (!jsonStr) continue;
        try {
          yield JSON.parse(jsonStr);
        } catch (err) {
          throw new HorizonsError(`invalid SSE payload: ${jsonStr}`);
        }
      }
    }
  }

  async listRegistered(): Promise<string[]> {
    return await this.client.request<string[]>("/api/v1/agents", { method: "GET" });
  }
}

File: horizons_py/horizons/agents.py
from __future__ import annotations

from typing import Any, AsyncIterator, Dict, List, Optional
from uuid import UUID

from .client import HorizonsClient
from . import models


class AgentsAPI:
    """Agent execution API methods."""

    def __init__(self, client: HorizonsClient) -> None:
        self._client = client

    async def run(
        self, *, agent_id: str, inputs: Optional[Dict[str, Any]] = None, project_id: Optional[UUID] = None
    ) -> models.AgentRunResult:
        """Run an agent once and return the typed run result."""
        body: Dict[str, Any] = {"agent_id": agent_id}
        if inputs is not None:
            body["inputs"] = inputs
        if project_id:
            body["project_id"] = str(project_id)
        resp = await self._client._request("POST", "/api/v1/agents/run", json=body)
        data = await self._client.json(resp)
        return models.AgentRunResult.model_validate(data["result"])

    async def chat_stream(
        self,
        *,
        agent_id: str,
        inputs: Optional[Dict[str, Any]] = None,
        project_id: Optional[UUID] = None,
    ) -> AsyncIterator[Dict[str, Any]]:
        """Stream agent chat events from `/api/v1/agents/chat` (SSE)."""
        body: Dict[str, Any] = {"agent_id": agent_id}
        if inputs is not None:
            body["inputs"] = inputs
        if project_id:
            body["project_id"] = str(project_id)
        async for event in self._client.sse("/api/v1/agents/chat", json=body):
            yield event

    async def list_registered(self) -> List[str]:
        """List registered agent identifiers for the current tenant scope."""
        resp = await self._client._request("GET", "/api/v1/agents")
        data = await self._client.json(resp)
        return list(data)
